### PR TITLE
Adding numpy_quaddtype package

### PR DESCRIPTION
**Pull request by SwayamInSync**
Checklist
- [ ] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [ ] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [ ] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [ ] Build number is 0.
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by SwayamInSync** [comment]
commenting to confirm my username as maintainer

File: recipes/numpy_quaddtype/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
+{% set name = "numpy_quaddtype" %}
+{% set version = "0.0.4" %}
+{% set python_req = "python >=3.10,<3.13" %}
```
Comment (ngoldbaum) [inline recipes/numpy_quaddtype/meta.yaml:3]: I think the complaint is coming from the version constraints here

Comment (SwayamInSync) [reply recipes/numpy_quaddtype/meta.yaml:3]: Yes, I set that intentionally as NumPy2.0 supports `3.9-3.12`

Comment (ngoldbaum) [reply recipes/numpy_quaddtype/meta.yaml:3]: @conda-forge/help-python any chance you can point Swayam at the right way to handle this?

I think all you need to do is depend on NumPy, you don't need to list any constraints that should be handled by that package. I'm not expert though, so let's see if any of the conda-forge maintainers want to help.

Comment (ocefpaf) [inline recipes/numpy_quaddtype/meta.yaml:3]:
```suggestion
```

File: recipes/numpy_quaddtype/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    - {{ stdlib("c") }}
+    - python
+    - pip
+    - meson
+    - meson-python
+    - setuptools
+    - sleef ==3.6.1
+    - numpy >=2.0
+  host:
+    - {{ python_req }}
```
Comment (ocefpaf) [inline recipes/numpy_quaddtype/meta.yaml:32]:
```suggestion
    - python
```

File: recipes/numpy_quaddtype/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    - pip
+    - setuptools
+    - meson
+    - meson-python
+    - sleef ==3.6.1
+    - numpy >=2.0
+  run:
+    - {{ python_req }}
```
Comment (ocefpaf) [inline recipes/numpy_quaddtype/meta.yaml:40]:
```suggestion
    - python
```

File: recipes/numpy_quaddtype/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    - {{ pin_compatible('sleef') }}
+    - {{ pin_compatible('numpy') }}
+
+test:
+  imports:
+    - numpy_quaddtype
+  requires:
+    - {{ python_req }}
```
Comment (ocefpaf) [inline recipes/numpy_quaddtype/meta.yaml:48]:
```suggestion
```

File: recipes/numpy_quaddtype/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    - numpy >=2.0
+    - pytest
+  source_files:
+    - quaddtype/tests
+  commands:
+    - pytest quaddtype/tests  # [not linux]
```
Comment (ocefpaf) [inline recipes/numpy_quaddtype/meta.yaml:54]: Why not on linux?

File: recipes/numpy_quaddtype/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    - pip
+    - setuptools
+    - meson
+    - meson-python
+    - sleef ==3.6.1
+    - numpy >=2.0
+  run:
+    - {{ python_req }}
+    - {{ pin_compatible('sleef') }}
+    - {{ pin_compatible('numpy') }}
```
Comment (ocefpaf) [inline recipes/numpy_quaddtype/meta.yaml:42]: If you are using numpy 2.0 you don't need this.
```suggestion
    - numpy
```

File: recipes/numpy_quaddtype/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+package:
+  name: {{ name }}
+  version: {{ version }}
+
+source:
+  url: https://github.com/SwayamInSync/numpy-user-dtypes/archive/refs/tags/quaddtype-v{{ version }}.tar.gz
+  sha256: a400f35cf980965486f9dd00c9d2289e877a4bd09a848f49b82de23e1855beea
+
+build:
+  number: 0
```
Comment (ocefpaf) [inline recipes/numpy_quaddtype/meta.yaml:14]:
```suggestion
  number: 0
  skip: True  # [py<310,py>=313]
```

File: recipes/numpy_quaddtype/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  script: |
+    cd quaddtype
+    {{ PYTHON }} -m pip install . -vv --no-build-isolation
```
Comment (ocefpaf) [inline recipes/numpy_quaddtype/meta.yaml:17]:
```suggestion
    {{ PYTHON }} -m pip install . -vv --no-deps --no-build-isolation
```

File: recipes/numpy_quaddtype/meta.yaml (hunk 5 above)
Comment (SwayamInSync) [reply recipes/numpy_quaddtype/meta.yaml:52]: Wierdly, one test was failing on linux so this is temporarily off
activating again

**Comment by SwayamInSync** [comment]
@ocefpaf I can see the artifacts are generated for all 3 OS, but in the logs I can't see the building and testing process.
Is this normal?

File: recipes/numpy_quaddtype/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  name: {{ name }}
+  version: {{ version }}
+
+source:
+  url: https://github.com/SwayamInSync/numpy-user-dtypes/archive/refs/tags/quaddtype-v{{ version }}.tar.gz
+  sha256: a400f35cf980965486f9dd00c9d2289e877a4bd09a848f49b82de23e1855beea
+
+build:
+  number: 0
+  skip: True  # [py<310,py>=313]
```
Comment (ocefpaf) [inline recipes/numpy_quaddtype/meta.yaml:14]:
```suggestion
  skip: True  # [py<310 or py>=313]
```

**Comment by SwayamInSync** [comment]
@ngoldbaum it seems the following test case of array addition is failing ONLY on Linux env (but passing on our GitHub and local linux testing with same Python version)

```python
def test_array_operations():
    arr1 = np.array(
        [QuadPrecision("1.5"), QuadPrecision("2.5"), QuadPrecision("3.5")])
    arr2 = np.array(
        [QuadPrecision("0.5"), QuadPrecision("1.0"), QuadPrecision("1.5")])

    result = arr1 + arr2
    expected = np.array(
        [QuadPrecision("2.0"), QuadPrecision("3.5"), QuadPrecision("5.0")])
    assert np.all(result == expected)
```
Will be doing some debugging on the conda's Linux environment to see where it is actually failing

**Comment by SwayamInSync** [comment]
This looks good now
cc: @ngoldbaum @ocefpaf 

PS: currently we are loading the source code from temporary (in development) repo, which will be replace in future. So it will possible by performing relevant edits on the feedstock right?

**Comment by ocefpaf** [comment]
> This looks good now cc: @ngoldbaum @ocefpaf
> 
> PS: currently we are loading the source code from temporary (in development) repo, which will be replace in future. So it will possible by performing relevant edits on the feedstock right?

As long as that is the current home for the project and the release is a stable one, we can merge this. You can change the source URLs later in the feedstock.

**Comment by SwayamInSync** [comment]
Thanks a lot for the help @ocefpaf 🙌🏻

**Pull request merged**